### PR TITLE
Drop IO-assumptions in processOHP

### PR DESCRIPTION
**Pull request by mlimbeck**

### CODE REVIEW

File: pkg/slayers/scion_spec.gobra [hunk 1]
```diff
@@ -486,6 +487,17 @@ pure func IsSupportedPkt(raw []byte) bool {
 		nextHdr != L4SCMP
 }
 
+ghost
+opaque
+decreases
+pure func IsSupportedPktSeq(raw seq[byte]) bool {
```
Comment (jcp19) [inline pkg/slayers/scion_spec.gobra:493]:
```suggestion
pure func IsSupportedRawPkt(raw seq[byte]) bool {
```

File: router/dataplane.go [hunk 2]
```diff
@@ -3664,11 +3652,12 @@ func (p *scionPacketProcessor) processOHP() (respr processResult, reserr error /
 	// for the rest of processing.
 	ohp.SecondHop.Mac = path.MAC(p.mac, ohp.Info, ohp.SecondHop, p.macBuffers.scionInput)
 	// @ fold ohp.SecondHop.Mem()
-	// @ fold s.Path.Mem(ubPath)
+	// @ fold acc(s.Path.Mem(ubPath), 1-R50)
+	// @ fold acc(p.scionLayer.Mem(ubScionL), 1-R15)
+	// assert reveal p.scionLayer.EqPathType(p.rawPkt)
```
Comment (jcp19) [inline router/dataplane.go:3657]:
```suggestion
```

File: router/dataplane.go [hunk 3]
```diff
@@ -3643,14 +3631,14 @@ func (p *scionPacketProcessor) processOHP() (respr processResult, reserr error /
...
 		// @ establishCannotRoute()
-		// @ defer fold p.scionLayer.Mem(ubScionL)
 		// @ fold p.d.validResult(processResult{}, false)
 		return processResult{}, serrors.WrapStr("bad source IA", cannotRoute,
 			"type", "ohp", "ingress", p.ingressID,
 			"neighborIA", neighborIA, "srcIA", s.SrcIA) /*@ , false, absReturnErr(processResult{}) @*/
 	}
-
-	// @ unfold s.Path.Mem(ubPath)
+	// assert reveal p.scionLayer.EqPathType(p.rawPkt)
```
Comment (jcp19) [inline router/dataplane.go:3639]:
```suggestion
```

File: router/dataplane.go [hunk 4]
```diff
@@ -3555,61 +3551,56 @@ func (p *scionPacketProcessor) processOHP() (respr processResult, reserr error /
...
+		// assert reveal p.scionLayer.EqPathType(p.rawPkt)
+		// @ unfold acc(p.scionLayer.Mem(ubScionL), 1-R15)
+		// @ unfold acc(s.Path.Mem(ubPath), 1-R50)
 		ohp.Info.UpdateSegID(ohp.FirstHop.Mac /*@, ohp.FirstHop.ToIO_HF() @*/)
-		// @ fold ohp.FirstHop.Mem()
-		// @ fold s.Path.Mem(ubPath)
-		// @ fold p.scionLayer.Mem(ubScionL)
+		// @ fold acc(s.Path.Mem(ubPath), 1-R50)
+		// @ fold acc(p.scionLayer.Mem(ubScionL), 1-R15)
+		// assert reveal p.scionLayer.EqPathType(p.rawPkt)
```
Comment (jcp19) [inline router/dataplane.go:3595]:
```suggestion
```

File: router/dataplane.go [hunk 5]
```diff
@@ -3555,61 +3551,56 @@ func (p *scionPacketProcessor) processOHP() (respr processResult, reserr error /
...
 		if compRes {
-			// @ defer fold p.scionLayer.Mem(ubScionL)
-			// @ defer fold s.Path.Mem(ubPath)
-			// @ defer fold ohp.FirstHop.Mem()
 			// TODO parameter problem -> invalid MAC
 			// @ fold p.d.validResult(processResult{}, false)
 			return processResult{}, serrors.New("MAC", "expected", fmt.Sprintf("%x", macCopy),
 				"actual", fmt.Sprintf("%x", ohp.FirstHop.Mac), "type", "ohp") /*@ , false, absReturnErr(processResult{}) @*/
 		}
+		// assert reveal p.scionLayer.EqPathType(p.rawPkt)
```
Comment (jcp19) [inline router/dataplane.go:3589]:
```suggestion
```

File: router/dataplane.go [hunk 6]
```diff
@@ -3555,61 +3551,56 @@ func (p *scionPacketProcessor) processOHP() (respr processResult, reserr error /
...
 			// @ fold p.d.validResult(processResult{}, false)
 			return processResult{}, err /*@ , false, absReturnErr(processResult{}) @*/
 		}
-		// @ unfold p.scionLayer.Mem(ubScionL)
-		// @ defer fold p.scionLayer.Mem(ubScionL)
-		// @ unfold s.Path.Mem(ubPath)
-		// @ defer fold s.Path.Mem(ubPath)
-		// @ unfold ohp.FirstHop.Mem()
-		// @ defer fold ohp.FirstHop.Mem()
+
```
Comment (jcp19) [inline router/dataplane.go:3603]:
```suggestion
```

File: router/dataplane.go [hunk 7]
```diff
@@ -3773,11 +3761,16 @@ func addEndhostPort(dst *net.IPAddr) (res *net.UDPAddr) {
...
+// @ requires  sl.Bytes(rawPkt, 0, len(rawPkt))
 // @ preserves buffer != nil && buffer.Mem()
-// @ preserves sl.Bytes(rawPkt, 0, len(rawPkt))
+// @ ensures   sl.Bytes(rawPkt, 0, len(rawPkt))
 // @ ensures   acc(s.Mem(rawPkt), R00)
 // @ ensures   res != nil ==> res.ErrorMem()
 // @ decreases
+// Contracts for IO-sepc
+// @ requires s.EqPathType(rawPkt)
+// @ requires !slayers.IsSupportedPkt(rawPkt)
```
Comment (jcp19) [inline router/dataplane.go:3772]: please put this together with the other preconditions (you may add a comment saying that these preconditions have to do with IO spec)

File: router/dataplane.go [hunk 8]
```diff
@@ -3788,10 +3781,13 @@ func updateSCIONLayer(rawPkt []byte, s *slayers.SCION, buffer gopacket.Serialize
 	if err := s.SerializeTo(buffer, gopacket.SerializeOptions{} /*@ , rawPkt @*/); err != nil {
 		return err
 	}
+
```
Comment (jcp19) [inline router/dataplane.go:3784]:
```suggestion
```

File: verification/utils/seqs/seqs.gobra [hunk 9]
```diff
@@ -36,4 +36,11 @@ requires size >= 0
...
-pure func NewSeqByteSlice(size int) (res seq[[]byte])
\ No newline at end of file
+pure func NewSeqByteSlice(size int) (res seq[[]byte])
+
+ghost
+requires forall i int :: { &ub[i] } 0 <= i && i < len(ub) ==> acc(&ub[i], _)
+ensures len(res) == len(ub)
+ensures forall i int :: { res[i] } 0 <= i && i < len(ub) ==> res[i] == ub[i]
+decreases _
+pure func ToSeqByte(ub []byte) (res seq[byte])
```
Comment (jcp19) [inline verification/utils/seqs/seqs.gobra:46]: if we only use this with `Slices`, why do we have the qps in the precondition?

**Review by jcp19 — CHANGES_REQUESTED** [review]

File: verification/utils/seqs/seqs.gobra (hunk 9 above)
Comment (mlimbeck) [reply verification/utils/seqs/seqs.gobra:49]: Fixed.

File: router/dataplane.go (hunk 7 above)
Comment (mlimbeck) [reply router/dataplane.go]: Can you double-check to see if it is better now?

**Review by jcp19 — APPROVED** [review]

**Pull request merged**